File: examples/python/crud.py
"""
AN5 example: generated Python client + runtime adapter.

Runnable against a reachable database via the AN5_DATABASE_URL environment
variable (postgres://... or Server=...;...). When the variable is unset the
script verifies that the generated client imports, registers metadata and can
be constructed — a compile/import smoke check.

Run:
    python -m compileall -q examples/python
    AN5_DATABASE_URL=postgres://u:p@host:5432/db python examples/python/crud.py
"""

import os
import sys

sys.path.insert(0, os.path.join(os.path.dirname(__file__), "..", "generated", "python"))
sys.path.insert(0, os.path.join(os.path.dirname(__file__), "..", "..", "an5Adapters", "python"))

import an5_client  # noqa: E402
import an5_metadata  # noqa: E402


def main():
    conn = os.getenv("AN5_DATABASE_URL")
    db = an5_client.An5Client(conn)

    if not conn:
        print("an5example Python client import check passed (AN5_DATABASE_URL not set)")
        return

    db.user.create(data={"email": "py@example.com", "name": "Py", "score": 1})
    u = db.user.find_first(where={"email": "py@example.com"})
    assert u and u["email"] == "py@example.com", "create/find via Python client"
    db.user.delete_many(where={"email": "py@example.com"})
    print("an5example Python CRUD smoke passed")


if __name__ == "__main__":
    main()

File: examples/typescript/crud.ts
/**
 * AN5 example: generated TypeScript types + @an5/adapters runtime against SQLite.
 *
 * Runnable offline (better-sqlite3 is a devDependency of an5example).
 *
 * Run: npm run build && node dist/examples/typescript/crud.js
 */
import { createAn5Adapter, setAdapterMetadata } from '@an5/adapters';
import { modelToTable, modelFields, relationMap } from '../../generated/typescript/an5Metadata';
import type { An5 } from '../../generated/typescript';

function ts(offsetDays = 0): string {
  const d = new Date(Date.now() + offsetDays * 86400000);
  return d.toISOString();
}

// SQLite-friendly metadata: strip "[dbo]." prefixes and brackets.
const sqliteTables: Record<string, string> = {};
for (const [model, table] of Object.entries(modelToTable)) {
  sqliteTables[model] = String(table).replace(/^\[dbo\]\./, '').replace(/^\[|\]$/g, '');
}
setAdapterMetadata({ modelToTable: sqliteTables, modelFields, relationMap });

interface ExampleDb {
  user: any;
  orders: any;
  order: any;
  $executeRawUnsafe(query: string, ...values: any[]): Promise<number>;
  $executeRaw(query: string, ...values: any[]): Promise<number>;
}

const db = createAn5Adapter({ connectionString: 'sqlite://:memory:' }) as unknown as ExampleDb & {
  $connect(): Promise<void>;
};

async function createTables(): Promise<void> {
  await db.$executeRaw(
    `CREATE TABLE IF NOT EXISTS users (
       id TEXT PRIMARY KEY,
       email TEXT NOT NULL UNIQUE,
       name TEXT NULL,
       isActive INTEGER NOT NULL DEFAULT 1,
       score INTEGER NOT NULL DEFAULT 0,
       createdAt TEXT NOT NULL DEFAULT (strftime('%Y-%m-%dT%H:%M:%fZ','now'))
     )`,
  );
  await db.$executeRaw(
    `CREATE TABLE IF NOT EXISTS orders (
       id TEXT PRIMARY KEY,
       userId TEXT NOT NULL,
       total INTEGER NOT NULL DEFAULT 0,
       status TEXT NULL,
       createdAt TEXT NOT NULL DEFAULT (strftime('%Y-%m-%dT%H:%M:%fZ','now')),
       FOREIGN KEY (userId) REFERENCES users(id)
     )`,
  );
}

async function main(): Promise<void> {
  await db.$connect();
  await createTables();

  const alice = await db.user.create({
    data: { email: 'alice@example.com', name: 'Alice', score: 123, createdAt: ts() },
    include: { orders: true, _count: true },
  });

  await db.orders.create({ data: { userId: alice.id, total: 250, status: 'open', createdAt: ts() } });
  await db.order.create({ data: { userId: alice.id, total: 75, status: 'open', createdAt: ts(1) } });

  const found = await db.user.findMany({ where: { email: { contains: 'alice' } }, include: { orders: true, _count: true } });
  console.log(`users matching 'alice': ${found.length}`);

  const order = await db.order.findFirst({ where: { id: (await db.orders.findMany({ take: 1 }))[0].id }, include: { user: true } });
  console.log(`order ${order?.id} belongs to ${order?.user?.email}`);

  const stats = await db.order.aggregate({ _sum: { total: true }, _count: true });
  console.log(`orders stat: ${JSON.stringify(stats)}`);

  await db.orders.deleteMany({ where: { status: 'open' } });
  await db.user.deleteMany({ where: { email: { contains: 'alice' } } });
  const remains = await db.user.count();
  console.log(`users after cleanup: ${remains}`);
}

main().then(
  () => console.log('an5example TypeScript CRUD example passed'),
  (err) => {
    console.error(err);
    throw err;
  },
);
